3.0.0 rc erikvk (#271)
Starcounter/Docs: docs/database-connection-string.md

## Connection string options

Name         | Type      | Required | Default value       | Description
:----------- | :-------- | :------- | :------------------ | :-----------------------------------------------------------------------
Database     | `string`  | Yes      | None                | A path string defining the database to use.
OpenMode     | `enum`    | No       | `CreateIfNotExists` | Specifies the database creation strategy.
StartMode    | `enum`    | No       | `StartIfNotRunning` | Specifies the database startup strategy.
StopMode     | `enum`    | No       | `IfWeStarted`       | Specifies the database shut down strategy.
ContextCount | `integer` | No       | `2` – `24`          | Specifies the number of database contexts allocated for this connection.

The options are further explored in the sections below.

### Database

The Database option is required in the connection string, and consists of an absolute or relative path to the directory where the Starcounter database files are located, or should be located if they do not yet exist.

> Internally, a Starcounter database is identified by it's physical location on disk and the version of Starcounter that is trying to access it. For example, attempting to start a database located at `/home/databases/TestDatabase` with Starcounter version `3.0.0-00001` will lead to this behavior:

Reviewer: >... it's physical location on disk ...

Should not it be `its`?

Also, the database is not identified by its physical location, but rather by the path provided in the connection string. If there would be a database located at `/path/db` and an environment variable called `path_db`, then two apps try to connect, one provides `/path/db` other one provides `%path_db%`. Starcounter will not detect that it is the same database, and will try to spin two `scdata` processes. Second attempt will fail, because the database files will be locked by the first `scdata` process.

Reviewer: It's already covered [here](https://github.com/Starcounter/Docs/pull/271/files#diff-cfb400e2cbd66e097e7a0f52bb00e03dR33), but the original description is a bit misleading, I think.

Author: > Should not it be its?

Yes it should!

Starcounter identifies database by the provided path, not by its physical location. Attempt to establish multiple connections to the same database using different paths will fail.

### OpenMode

The OpenMode option defines how to interpret the database path on connection. It can have the following string values:

Value               | Description
:------------------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
`Open`              | Opens the database directory from the path specified in the Database option, if it exist. If the directory does not exist, or if there is no Starcounter database located there, an exception is thrown.
`CreateIfNotExists` | Opens the database directory from the path specified in the Database option, if it exist. If the directory does not exist, or if there is no Starcounter database located there, it will create the necessary directories and database files.

### StartMode

> The StartMode option defines how to start and/or connect to the database process once the database files have been located and/or created.

Reviewer: ```suggestion
The `StartMode` option defines how to start and/or connect to the database process once the database files have been located and/or created.
```

Author: OK, let's use syntax quotes for connection string options. I didn't want to over-use it, but sure :)

Author: (no need to add suggestions for it, I'll fix it)

Reviewer: I prefer to use syntax quotes for code pieces to avoid unnecessary grammar checks. StartMode is not a correct word according to English grammar, but when it's `StartMode`, it's easier to filter out.

My personal preference.

Author: I don't think grammar checks is a good enough reason to use syntax quotes. It should be used for syntactic constructions that represent some formal notion in a system. I think these options qualify, so I'll change it.

Value               | Description
:------------------ | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
`RequireStarted`    | Expects the database process to already be running and will connect to it if it is, or throw an exception if it is not.
`StartIfNotRunning` | Connects to the database process if it is already running, or starts and connects to it if it is not.

> `StartExclusive`    | Expects no database process to be already running, and starts and connects to a new process if that is the case. If a database process is already running, an exception is thrown.

Reviewer: ```suggestion
`StartExclusive`    | Expects no database process to be already running, starts and connects to a new process if that is the case. If a database process is already running, an exception is thrown.
```

Author: This is a tricky use of punctuation. Your edit actually makes the sentence mean something different, since the "and" you're removing is not used to end an enumeration but for joining two propositions. Regardless, I think I'll just rewrite it so it's not so weird.

### StopMode

> The StopMode option defines whether to stop the database process once the connection to it is closed, for example when an application is terminated.

Reviewer: ```suggestion
The `StopMode` option defines whether to stop the database process once the connection to it is closed, for example when an application is terminated.
```

Reviewer: Currently we have an issue, that an unexpected termination of an application leaves `scdata` process running. Works fine if the app shuts down properly.

Author: We should regard that as a bug, right? In which case it doesn't really change the intended semantics for `StopMode`

Reviewer: Yes, just telling you in case you would want to test it.

Value         | Description
:------------ | :-----------------------------------------------------------------------------------
`IfWeStarted` | Stops the database process if and only if it was started by the current application.
`Never`       | Leaves the database process running, regardless of how it was started.

**Note:** the `Never` option is not effective when the application process is forcibly killed, for example from a task manager or by `Ctrl + C`. In this case operating system will kill the corresponding database process as well.

### ContextCount

The ContextCount option defines the number of **database contexts** to use in the database connection.

> "Database context", as used here, should not be confused with `IDatabaseContext` as used in the .NET API for transactions. In the context of database connections, database contexts refer to the entrypoints to the database from the applications' point of view. When an application thread is assigned a database context, it can read and write to the database.

Reviewer: When an application thread has a database context assigned, it can read and write to the database.

The same database context may be used between multiple applications threads, for example with the `async`/`await`. Starcounter manages this automatically.

Author: The automatic management you mention is implemented by means of a custom synchronization context, right? Or is that just for transactions? Or are they the same notion?

Reviewer: It's the same thing, done through the custom synchronization context.

Each Starcounter database has a maximum of 31 database contexts available for allocation between multiple application processes. By default, each Starcounter application tries to allocate twice as many contexts as CPU cores available, with a minimum of 2 and a maximum of 24.

The ideal configuration is to have a database context per CPU core. Having more database contexts than CPU cores might increase performance in certain concurrent database access workloads. Having fewer database contexts than CPU cores might decrease performance of concurrent database access during high concurrent load.

Use `ContextCount` option to manually adjust the number of database contexts to occupy in this connection.

## Example

```text
Database=./.database/Sample;OpenMode=CreateIfNotExists;StartMode=StartIfNotRunning;StopMode=IfWeStarted;ContextCount=10
```

> This string defines the database option using a relative path to a database `Sample`, located in the `.database` directory of the current directory. If such a directory does not exist at the time of connection, the OpenMode option of `CreateIfNotExists` defines that it should be created. If not running already, the StartMode option of `StartIfNotRunning` defines that it should be started. When te app disconnects from the database, it will stop the database process if it started, due to the StopMode option being defined as `IfWeStarted`. Lastly, we set the number of database contexts to 10.

Author: ```suggestion
This string defines the database option using a relative path to a database `Sample`, located in the `.database` folder in the current directory.
```

> This string defines the database option using a relative path to a database `Sample`, located in the `.database` directory of the current directory. If such a directory does not exist at the time of connection, the OpenMode option of `CreateIfNotExists` defines that it should be created. If not running already, the StartMode option of `StartIfNotRunning` defines that it should be started. When te app disconnects from the database, it will stop the database process if it started, due to the StopMode option being defined as `IfWeStarted`. Lastly, we set the number of database contexts to 10.

Author: ```suggestion
This string defines the database option using a relative path to a database `Sample`, located in the `.database` folder in the current directory. If such a directory does not exist at the time of connection, the OpenMode option of `CreateIfNotExists` defines that it should be created. If not running already, the StartMode option of `StartIfNotRunning` defines that it should be started. When te app disconnects from the database, it will stop the database process if it started, due to the StopMode option being defined as `IfWeStarted`. Lastly, we set the number of database contexts to 10.
```

> This string defines the database option using a relative path to a database `Sample`, located in the `.database` directory of the current directory. If such a directory does not exist at the time of connection, the OpenMode option of `CreateIfNotExists` defines that it should be created. If not running already, the StartMode option of `StartIfNotRunning` defines that it should be started. When te app disconnects from the database, it will stop the database process if it started, due to the StopMode option being defined as `IfWeStarted`. Lastly, we set the number of database contexts to 10.

Reviewer: ```suggestion
This string defines the database option using a relative path to a database `Sample`, located in the `.database` directory of the current directory. If such a directory does not exist at the time of connection, the `OpenMode` option of `CreateIfNotExists` defines that it should be created. If not running already, the StartMode option of `StartIfNotRunning` defines that it should be started. When the app disconnects from the database, it will stop the database process if it started, due to the `StopMode` option being defined as `IfWeStarted`. Lastly, we set the number of database contexts to 10.
```
